src/Microsoft.Identity.Client/Internal/ClientCredentialWrapper.cs

```diff
@@ -17,41 +24,154 @@ namespace Microsoft.Identity.Client.Internal
+        #endregion TestConstructors
+        private void CheckCertificateKeySize(X509Certificate2 cert)
+        {
+#if DESKTOP
+            if (cert.PublicKey.Key.KeySize < MinKeySizeInBits)
+            {
+                throw new ArgumentOutOfRangeException(nameof(cert),
+                    string.Format(CultureInfo.InvariantCulture, MsalErrorMessage.CertificateKeySizeTooSmallTemplate,
+                        MinKeySizeInBits));
+            }
+#endif
+        }
 
-            Secret = secret;
+        private void ValidateCredentialParameters(ApplicationConfiguration config)
+        {
+            int countOfCredentialTypesSpecified = 0;
+
+            if (!string.IsNullOrWhiteSpace(config.ClientSecret))
+            {
+                countOfCredentialTypesSpecified++;
+            }
+
+            if (config.ClientCredentialCertificate != null)
+            {
+                countOfCredentialTypesSpecified++;
+            }
+
+            if (!string.IsNullOrWhiteSpace(config.SignedClientAssertion))
+            {
+                countOfCredentialTypesSpecified++;
+            }
+
+            if (countOfCredentialTypesSpecified > 1)
+            {
+                throw new MsalClientException(MsalError.ClientCredentialAuthenticationTypesAreMutuallyExclusive, MsalErrorMessage.ClientCredentialAuthenticationTypesAreMutuallyExclusive);
+            }
+
+            if (!string.IsNullOrWhiteSpace(config.ClientSecret))
+            {
+                AuthenticationType = ConfidentialClientAuthenticationType.ClientSecret;
+            }
+
+            if (config.ClientCredentialCertificate != null)
+            {
+                if (config.ClaimsToSign != null && config.ClaimsToSign.Any())
+                {
+                    AuthenticationType = ConfidentialClientAuthenticationType.ClientCertificateWithClaims;
+                }
+                else
+                {
+                    AuthenticationType = ConfidentialClientAuthenticationType.ClientCertificate;
+                }
+            }
+
+            if (!string.IsNullOrWhiteSpace(config.SignedClientAssertion))
+            {
+                AuthenticationType = ConfidentialClientAuthenticationType.SignedClientAssertion;
+            }
         }
 
+        internal byte[] Sign(ICryptographyManager cryptographyManager, string message)
+        {
+            return cryptographyManager.SignWithCertificate(message, Certificate);
+        }
+
+        private static readonly int s_minKeySizeInBits = 2048;
+        public static int MinKeySizeInBits { get { return s_minKeySizeInBits; } }
+        internal string Thumbprint { get { return Base64UrlHelpers.Encode(Certificate.GetCertHash()); } }
+        internal X509Certificate2 Certificate { get; }
+        // The cached assertion created from the JWT signing operation
+        internal string CachedAssertion { get; set; }
+        internal long ValidTo { get; set; }
+        internal bool ContainsX5C { get; set; }
```

Comment:
Can we make all setters private (or readonly if possible)

Reply:
+1